### PR TITLE
Make a generic type out of Registry

### PR DESCRIPTION
**Pull request by jrapin**
## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Improves typing since so far registry were returning Any.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/common/decorators.py [hunk 1]
```diff
@@ -3,28 +3,36 @@
...
 
-class Registry(dict):  # type: ignore
+X = TypeVar("X")
+
+
+# pylint does not understand Dict, and mypy needs Dict to be generic...
+# the following is (very) hacky but solves both issues for now
+# It can be safely reverted to Registry(dict) though...
+# pylint: disable=too-many-ancestors
+class Registry(Dict[str, X], UserDict):  # type: ignore
```
Comment (lerks) [inline nevergrad/common/decorators.py]: I don't think this does what you want. Dict means this is a *real* dict, whereas UserDict means that this has a dict as the `data` attribute. They define exactly the same methods, so the one that comes in the MRO overrides the second one. What this means here is that UserDict has no effect at all. If you want to keep UserDict make the first a MutableMapping.

File: nevergrad/common/decorators.py [hunk 2]
```diff
@@ -3,28 +3,36 @@
...
         self._information: Dict[str, Dict[Any, Any]] = {}
 
-    def register(self, obj: Any, info: Optional[Dict[Any, Any]] = None) -> Any:
+    def register(self, obj: X, info: Optional[Dict[Any, Any]] = None) -> Any:
         """Decorator method for registering functions/classes
         The info variable can be filled up using the register_with_info
         decorator instead of this one.
         """
-        name = obj.__name__ if hasattr(obj, "__name__") else obj.__class__.__name__
+        name = obj.__name__ if hasattr(obj, "__name__") else obj.__class__.__name__  # type: ignore
```
Comment (lerks) [inline nevergrad/common/decorators.py]: This is `getattr(obj, "__name__", obj.__class__.__name__)`.

File: nevergrad/common/decorators.py [hunk 3]
```diff
@@ -3,28 +3,36 @@
...
+class Registry(Dict[str, X], UserDict):  # type: ignore
     """Registers function or classes as a dict.
     """
 
     def __init__(self) -> None:
         super().__init__()
         self._information: Dict[str, Dict[Any, Any]] = {}
 
-    def register(self, obj: Any, info: Optional[Dict[Any, Any]] = None) -> Any:
+    def register(self, obj: X, info: Optional[Dict[Any, Any]] = None) -> Any:
```
Comment (lerks) [inline nevergrad/common/decorators.py]: `register` is now also *returning* an object of type `X`

File: nevergrad/benchmark/experiments.py [hunk 4]
```diff
@@ -257,9 +257,11 @@ def mldaas(seed: Optional[int] = None) -> Iterator[Experiment]:
                             yield xp
 
 
+# def arcoating(seed: Optional[int] = None) -> Iterator[Experiment]:
 @registry.register
-def arcoating(seed: Optional[int] = None) -> Iterator[Experiment]:
+def arcoating() -> Iterator[Experiment]:
     func = ARCoating()
+    seed = 12
```
Comment (lerks) [inline nevergrad/benchmark/experiments.py]: Why do you need this? Either this isn't used, or you're introducing a seemingly unrelated functional change.

File: nevergrad/common/decorators.py (hunk 3 above)
Comment (jrapin) [reply nevergrad/common/decorators.py]: Good point

File: nevergrad/common/decorators.py (hunk 2 above)
Comment (jrapin) [reply nevergrad/common/decorators.py]: Oh did not know there was a default!

File: nevergrad/common/decorators.py (hunk 1 above)
Comment (jrapin) [reply nevergrad/common/decorators.py]: It actually *does* what I want but it is really hacky:
- the methods which are used are the ones in `Dict[str, X]`, which is convenient since they are typed.
- `UserDict` is basically only there so that pylint does not complain since it does not recognize Dict. The `data` field is unused... actually it does not even exist, not sure why :s
- using `Registry(UserDict, MutableMapping[str, X])` has untyped getitem... I would have to rewrite all dict methods :s

File: nevergrad/benchmark/experiments.py (hunk 4 above)
Comment (jrapin) [reply nevergrad/benchmark/experiments.py]: This was just a test that I removed probably a second after your comment

**Comment by jrapin** [comment]
I can't make it work in a non-hacky way... is it worth it ? :s

File: nevergrad/common/decorators.py (hunk 1 above)
Comment (lerks) [reply nevergrad/common/decorators.py]: You can have this extend just MutableMapping, remove all dict or UserDict superclass, then manually define a `data` attribute, which is a dict, and manually implement the 5 abstract methods of MutableMapping (see there [here](https://docs.python.org/3/library/collections.abc.html)). These methods will basically just be
```py
def foo(*args, *kwargs):
    return self.data.foo(*args, **kwargs)
```
Once you do so you get for free all other methods of the dictionary interface. So, in short, your class will be a regular class (not a dict) but will be dict-like.

Comment (jrapin) [reply nevergrad/common/decorators.py]: I was a bit afraid pylint would not recognize it either, but it worked perfectly! thank you!
